src/ast/mod.rs

```diff
@@ -893,26 +1030,44 @@ where
 
 impl<Head, Leaf, Ann> GenericFact<Head, Leaf, Ann>
 where
-    Ann: Clone,
+    Ann: Clone + Default,
     Head: Clone + Display,
     Leaf: Clone + PartialEq + Eq + Display + Hash,
 {
-    pub(crate) fn map_exprs<Head2, Leaf2>(
-        &self,
-        f: &mut impl FnMut(&GenericExpr<Head, Leaf, Ann>) -> GenericExpr<Head2, Leaf2, Ann>,
-    ) -> GenericFact<Head2, Leaf2, Ann> {
+    pub(crate) fn map_exprs(
+        self,
+        f: &mut impl FnMut(GenericExpr<Head, Leaf, Ann>) -> GenericExpr<Head, Leaf, Ann>,
+    ) -> GenericFact<Head, Leaf, Ann> {
         match self {
-            GenericFact::Eq(exprs) => GenericFact::Eq(exprs.iter().map(f).collect()),
-            GenericFact::Fact(expr) => GenericFact::Fact(f(expr)),
+            GenericFact::Eq(exprs) => {
+                GenericFact::Eq(exprs.into_iter().map(|expr| expr.map(f)).collect())
+            }
+            GenericFact::Fact(expr) => GenericFact::Fact(expr.map(f)),
         }
     }
 
-    pub(crate) fn subst<Leaf2, Head2>(
```

Comment:
Same question here. The original `subst` method seems more powerful, and it is very common to substitute a leaf (variable) with an expression (this is how the substitution works for lambda calculus), while map_leafs is more like renaming than substitution. Could we still do what the original `subst` did in the new proposed interface?

Reply:
added old subst